File: leetcode-in-python3/solution/2_add_two_numbers.py
from datastruct.linknode import ListNode
import unittest


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        carry = 0
        head = ListNode(0)
        node = head
        while l1 is not None or l2 is not None or carry != 0:
            sum_ = carry
            if l1 is not None:
                sum_ += l1.val
                l1 = l1.next
            if l2 is not None:
                sum_ += l2.val
                l2 = l2.next

            carry = sum_ // 10
            node.next = ListNode(sum_ % 10)
            node = node.next
        return head.next


class TestSolution(unittest.TestCase):

    def setUp(self):
        self.s = Solution()

    def test_normal(self):
        l1 = [2, 4, 3]
        l2 = [5, 6, 4]
        answer = [7, 0, 8]
        l1 = ListNode.create(l1)
        l2 = ListNode.create(l2)
        ans = self.s.addTwoNumbers(l1, l2)
        self.assertEqual(answer,ans .to_list())

    def test_99999(self):
        l1 = [9, 9, 9, 9, 9, 9, 9]
        l2 = [9, 9, 9, 9]
        answer = [8, 9, 9, 9, 0, 0, 0, 1]
        l1 = ListNode.create(l1)
        l2 = ListNode.create(l2)
        ans = self.s.addTwoNumbers(l1, l2)
        self.assertEqual(answer,ans .to_list())


if __name__ == '__main__':
    unittest.main()

File: leetcode-in-rust/src/solution/_2_add_two_numbers.rs
use crate::datastruct::linknode::ListNode;

pub struct Solution;

impl Solution {
    pub fn add_two_numbers(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let (mut l1, mut l2) = (l1, l2);

        let mut carry = 0;
        let mut head = ListNode::new(0);
        let mut node = &mut head;

        while l1.is_some() || l2.is_some() || carry != 0 {
            let mut sum = carry;

            if let Some(list_node) = l1 {
                sum += list_node.val;
                l1 = list_node.next
            }

            if let Some(list_node) = l2 {
                sum += list_node.val;
                l2 = list_node.next;
            }
            carry = sum / 10;
            node.next = Some(Box::new(ListNode::new(sum % 10)));
            node = node.next.as_mut().unwrap();
        }
        head.next
    }
}

#[cfg(test)]
mod tests {
    use crate::datastruct::linknode::ListNode;
    use super::*;

    #[test]
    fn test_solution() {
        let test_case = vec![
            (vec![2, 4, 3], vec![5, 6, 4], vec![7, 0, 8]),
            (vec![9, 9, 9, 9, 9, 9, 9], vec![9, 9, 9, 9], vec![8, 9, 9, 9, 0, 0, 0, 1]),
        ];

        for (num1, num2, num3) in test_case {
            let l1 = ListNode::create(num1);
            let l2 = ListNode::create(num2);
            let answer = ListNode::create(num3);
            let ans = Solution::add_two_numbers(l1, l2);
            assert_eq!(ans, answer)
        }
    }
}
